### PR TITLE
Storages: Change CompressionCodecPtr from shared_ptr to unique_ptr

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #8789

A `CompressionCodec` object will not be shared, so `CompressionCodecPtr` is not necessary to be a type of `shared_ptr`, but `unique_ptr` is enough. 

### What is changed and how it works?

Change the type of `CompressionCodecPtr` from `shared_ptr` to `unique_ptr`.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
We can use one CompressionCodec to create multiple Read/WriteBuffer?

File: dbms/src/IO/Compression/CompressionFactory.h [hunk 1]
```diff
@@ -74,17 +74,8 @@ class CompressionFactory
...
-            codecs.reserve(settings.settings.size());
-            for (const auto & setting : settings.settings)
-            {
-                codecs.push_back(create(setting));
-            }
-            return std::make_shared<CompressionCodecMultiple>(std::move(codecs));
-        }
-        return create(settings.settings.front());
+        return settings.settings.size() > 1 ? std::make_unique<CompressionCodecMultiple>(createCodecs(settings))
+                                            : create(settings.settings.front());
```
Comment (JaySon-Huang) [inline dbms/src/IO/Compression/CompressionFactory.h:78]: check that the size of `settings` must not be zero is better. Or this line may crash

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Review by JaySon-Huang — APPROVED** [review]
lgtm

**Pull request merged**